tar/lib/src/stream.dart

```diff
@@ -0,0 +1,135 @@
+// Copyright 2020
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.import 'constants.dart';
+
+import 'dart:async';
+import 'sparse_entry.dart';
+
+/// Generates a stream of the sparse file contents of size [size], given
+/// [sparseHoles] and the raw content in [source].
+Stream<int> sparseStream(
+    Stream<int> source, List<SparseEntry> sparseHoles, int size) async* {
+  ArgumentError.checkNotNull(source, 'source');
+  ArgumentError.checkNotNull(sparseHoles, 'sparseHoles');
+  ArgumentError.checkNotNull(size, 'size');
+
+  /// Current logical position in sparse file, including the bytes in
+  /// [partialChunk].
+  var position = 0;
+
+  /// Index of the next sparse hole in [sparseHoles] to be processed.
+  var sparseHoleIndex = 0;
+
+  /// Iterator through [source] to obtain the data bytes.
+  final iterator = StreamIterator(source);
+
+  while (position < size) {
+    /// Yield all the necessary sparse holes.
+    while (sparseHoleIndex < sparseHoles.length) {
+      final sparseHole = sparseHoles[sparseHoleIndex];
+      if (sparseHole.offset == position) {
+        for (final byte in List<int>.filled(sparseHole.length, 0)) {
+          yield byte;
+        }
+        position += sparseHole.length;
+        sparseHoleIndex++;
+      } else {
+        break;
+      }
+    }
+
+    if (position == size) break;
+
+    /// Grab a new chunk from [source].
+    if (!await iterator.moveNext()) {
+      throw StateError('sparseStream size is larger than what the stream '
+          'supports!');
+    }
+    yield iterator.current;
+    position++;
+  }
+}
+
+/// Implementation of a Stream iterator that reads a chunked stream of [List<T>].
+class ChunkedStreamIterator<T> {
+  /// Underlying iterator that iterates through the original stream.
+  final StreamIterator<T> _iterator;
+
+  /// Keeps track of the number of elements left in the current substream.
+  int _toRead = 0;
+
+  /// Keeps track of the current substream we are supporting.
+  int _substreamId = 0;
+
+  ChunkedStreamIterator(Stream<List<T>> stream)
+      : _iterator = StreamIterator(stream.expand((element) => element));
```

Comment:
Thanks for the suggestion @isoos ! 

I think the primary goal we're trying to achieve is being able to read the input stream just once, without having to buffer the full contents of the file if we didn't want it. The use cases for this functionality are for example when we wish to only process certain file entries within a tar file while skipping the rest based on the headers, which allows us to save on memory.

You are right in that the current `ChunkedStreamIterator` as implemented here is inefficient, and I'm currently looking at removing the `stream.expand()` step and processing the stream in chunks, which should hopefully improve the space and time efficiency, and afterwards changing the constructor to a factory constructor that might return a more optimized iterator for `Uint8List`s by taking advantage of the `sublistView`.

I just took a look at `package:buffer`, and I'm not sure how I would use `package:buffer` to its fullest potential here. This `ChunkedStreamIterator` is really only meant to read and pass back chunks of raw bytes on demand. Could you advise me on how I might use the package to achieve the above goals?

/cc @jonasfj

Reply:
> I think the primary goal we're trying to achieve is being able to read the input stream just once, without having to buffer the full contents of the file if we didn't want it.

Streams will buffer the events by default, as you have no control on when the Dart VM reads and forwards the bytes (or in how large chunks). The byte-stream coming from a `File` is an efficient 1-byte per value `Uint8List`, and leaving it in the Stream or buffering it via `ByteDataReader` is roughly the same overhead. However, `Stream.expand()` will blow up the memory requirement, as it is no longer a 1-byte `int` value, rather a 8-byte int value, and each value is copied to an expanded memory buffer.

For controlled byte-streams from files (and skipping ranges of bytes efficiently), you will need to use `RandomAccessFile`.

Reply:
> Streams will buffer the events by default, as you have no control on when the Dart VM reads and forwards the bytes (or in how large chunks). 

hm, to my understanding, `File`'s `openRead()` is a single-subscription stream, and though I don't have precise control over when the Dart VM reads and forwards the bytes, wouldn't wrapping it in a `StreamIterator` pause the stream once the next element is processed (i.e. at the end of every `moveNext()`?

Reply:
Not sure, maybe calling the `StreamSubscription.pause` and `.resume` in this case the VM does the right thing. We should test it with e.g. a 1-GB .tar file and see how the memory allocation profile goes.

Reply:
Sorry about the delay, just got around to testing it. The `expand` really did make it super inefficient, so I modified the iterator to deal with chunks rather than expanding the elements out.

The testing was done on a 1-GB .tar file comprising two smaller files of 500 MB each, and the code used for testing can be found [here](https://gist.github.com/walnutdust/3fd79f9be2c0e324ba683d0bc93beff7). The testing code comprises of two parts, one where I ran through the file with the `TarReader` implemented with the `ChunkedStreamIterator` implementation, and another where we just did an `await for` loop through the elements of the file `Stream` doing no work to it. Time measured was the average of 100 iterations.

|Metric| TarReader | File Stream|
|----------|-----------|-----------|
|Average function run time| 341145.94 us | 337024.06 us |
|Peak Process Memory| 338.4MB | 328.7MB|
|New Generation Used:|1.5MB of 32.0MB | 1.5MB of 32.0MB|
|New Generation External| 21.5MB | 21.7MB|
|New Generation Collections| 748 | 748 |
|New Generation Average Collection Time| 0.37 ms| 0.35 ms|
|Old Generation Used| 6.4MB of 7.3MB| 5.7MB of 6.3MB|
|Old Generation External| 64B| 64B|
|Old Generation Collections| 4|3|
|Old Generation Average Collection Time| 1.78 ms| 2.46 ms|
|Total Generation Used| 7.9MB of 39.3MB| 7.2MB of 38.3MB|
|Total Generation External| 21.5MB| 21.7MB|
|Total Generation Collections| 752| 751|
|Total Generation Average Collection Time| 0.38 ms| 0.36 ms|

Naively looking at this, it appears to me that the `ChunkedStreamIterator` achieves our intended goals, and is pretty efficient about it. I am quite new to profiling, so please feel free to let me know if you want me to run more experiments with more parameters, or if there's something else that I should be analyzing!